internal/handlers/jsonb1/msg_aggregate.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jsonb1
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/FerretDB/FerretDB/internal/pg"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/wire"
+)
+
+// MsgAggregate executes an aggregation pipeline on documents in a collection.
+func (h *storage) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, error) {
+	document, err := msg.Document()
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	m := document.Map()
+	pipeline := m["pipeline"].(types.Array)
+	db := m["$db"].(string)
+	collection := m["aggregate"].(string)
+
+	sql, args, err := aggregateToSQL(pipeline, db, collection)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	rows, err := h.pgPool.Query(ctx, sql, args...)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	defer rows.Close()
+
+	var docs types.Array
+
+	for {
+		doc, err := nextRow(rows)
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+		if doc == nil {
+			break
+		}
+
+		docs = append(docs, *doc)
+	}
+
+	var reply wire.OpMsg
+	err = reply.SetSections(wire.OpMsgSection{
+		Documents: []types.Document{types.MustMakeDocument(
+			"cursor", types.MustMakeDocument(
+				"firstBatch", docs,
+				"id", int64(0), // TODO
+				"ns", db+"."+collection,
+			),
+			"ok", float64(1),
+		)},
+	})
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	return &reply, nil
+}
+
+func aggregateToSQL(pipeline types.Array, db, collection string) (string, []interface{}, error) {
+	filterPairs := []interface{}{}
+	groupByFields := map[string]string{}
+	sortByFields := map[string]interface{}{}
+	aggregations := map[string]string{}
+	for _, item := range pipeline {
+		if p, ok := item.(types.Document); ok {
+			if p.Map()["$match"] != nil {
```

Comment:
Only on `$match`, `$sort`, `$group` can be present, right? If yes, let's replace it with `switch p.Command()`